lib/graphql/batch/execution_strategy.rb

```diff
@@ -1,57 +1,62 @@
 module GraphQL::Batch
   class ExecutionStrategy < GraphQL::Query::SerialExecution
-    attr_reader :batched_queries
-
-    def initialize
-      @batched_queries = Hash.new{ |hash, key| hash[key] = [] }
-    end
-
     class OperationResolution < GraphQL::Query::SerialExecution::OperationResolution
       def result
-        result = super
-        until execution_strategy.batched_queries.empty?
-          queries = execution_strategy.batched_queries.shift.last
-          queries.first.class.execute(queries)
-        end
-        result
+        GraphQL::Batch.wait(super)
+      ensure
+        GraphQL::Batch::Executor.current.clear
       end
     end
 
     class SelectionResolution < GraphQL::Query::SerialExecution::SelectionResolution
       def result
-        result_hash = super
-        result_hash.each do |key, value|
-          if value.is_a?(FieldResolution)
-            value.result_hash = result_hash
+        lazy_resolve_object(super)
+      end
+
+      private
+
+      def lazy_resolve_object(obj)
+        queries = nil
+        case obj
+        when Array
+          array = obj
+          array.each_with_index do |value, i|
+            if query = lazy_key_assign(array, i, value)
+              (queries ||= []) << query
```

Comment:
I could add benchmarks, but I don't think that is a high priority right now, since a contrived benchmarks can be quite misleading.

I'm not crazy for considering performance considering that liquid and identity cache have generally been at the top of our CPU profiles in production, and removing object allocation has been the primary way of improving those code paths.